docs/debugger/runtime-debugging.md

```diff
@@ -1,36 +1,113 @@
+- IntelliJ
+- IGV - with [extension described herein](../../tools/enso4igv/IGV.md)
+- [NetBeans](http://netbeans.apache.org) - with the same
+  [extension as IGV](../../tools/enso4igv/IGV.md) is using
+
+There are two ways to start JVM in a debug mode. The first one is fully
+integrated into the `sbt` build system. It builds the engine (in case there were
+any changes) and then starts the JVM in debug mode trying to attach to port
+`5005`:
 
 ```sh
 sbt:enso> runEngineDistribution --debug --run ./test/Tests/src/Data/Numbers_Spec.enso
 ```
 
-The second options is to pass in special JVM arguments when launching the
-`bin/enso` from the engine distribution:
+The second options gives one a complete control as it launches everything from a
+command line. By specifying `JAVA_OPTS` environment variable one influences the
+special JVM arguments when launching the `bin/enso` from the engine
+distribution:
 
 ```bash
 enso$ JAVA_OPTS=-agentlib:jdwp=transport=dt_socket,server=n,address=5005 ./built-distribution/enso-engine-*/enso-*/bin/enso --run ./test/Tests/src/Data/Numbers_Spec.enso
 ```
 
-### Tips and tricks
+Both of the approaches launch the JVM in a _debug mode_. Once the JVM is
+started, simply attach the debugger to the JVM process at the specified port
+(usually `5005` is used).
+
+### Attaching from VSCode
+
+First of all make sure your VSCode contains the
+[Enso supporting extension](https://marketplace.visualstudio.com/items?itemName=Enso.enso4vscode).
+Follow the [instructions provided herein](../../tools/enso4igv/README.md) to
+install everything that's needed.
+
+Once the
+[Enso extension](https://marketplace.visualstudio.com/items?itemName=Enso.enso4vscode)
+is installed, open the root of Enso repository as workspace and select _"Listen
+to 5005"_ debug configuration:
+
+![Listen to 5005](https://github.com/enso-org/enso/assets/26887752/1874bcb1-cf8b-4df4-92d8-e7fb57e1b17a)
+
+Once the connection with the Enso JVM is made, debug the Java code as usual.
+
+### Mapping Java and Enso Code
+
+The _Enso interpreter_ maintains a mapping between its Java code and appropriate
+_Enso Code_ internally. It shall be no problem to use that information in the
+debugger. All the mentioned IDEs (alas except IntelliJ) support such _mixed
+debugging_ of Java and Enso code.
+
+Put a breakpoint into `.enso` file and after a while the breakpoint is hit and
+one can inspect variables, step over the _Enso statements_ and more...
+
+![Breakpoint in Enso](https://github.com/enso-org/enso/assets/26887752/54ae4126-f77a-4463-9647-4dd3a5f83526)
+
+...as one can seamlessly switch to debugging on the Enso interpreter itself! One
+can place breakpoint into Java class like `PanicException.java` and continue
+debugging with `F5`:
+
+![Breakpoint in Java](https://github.com/enso-org/enso/assets/26887752/db3fbe4e-3bb3-4d4a-bb2a-b5039f716c85)
+
+Should one ever want to jump back from Java to Enso one can use the _"Pause in
```

Comment:
If I already have breakpoints set in Enso code, do I still need to use "Pause in GraalVM Script", or will the debugger stop at the breakpoints without it?

Reply:
Your mileage may vary. The "Pause in GraalVM Script' is using a (checksum) heuristic to find the script code. If subsequent executions of the code can find the _same breakpoint_, then it is enough to put the breakpoint into Enso code.

If we get reproducible test cases when the breakpoint isn't hit, we can investigates what's the cause and fix it. The devil is in the detail: if we have enough details, we can improve your experience.

Reply:
So Pause is definitely needed if an Enso breakpoint hasn't been hit yet?

Reply:
Until a breakpoint is hit, the debuggee runs. That's the common behavior of all the debuggers, I think.